### PR TITLE
Fix potential crash if using UTF8 directory names.

### PR DESCRIPTION
**Pull request by kjnilsson**

### CODE REVIEW

File: src/osiris_log.erl [hunk 1]
```diff
@@ -1903,7 +1903,7 @@ update_retention(Retention,
 evaluate_retention(Dir, Specs) when is_list(Dir) ->
     % convert to binary for faster operations later
     % mostly in segment_from_index_file/1
-    evaluate_retention(list_to_binary(Dir), Specs);
+    evaluate_retention(filename:join(Dir, <<"">>), Specs);
```
Comment (michaelklishin) [inline src/osiris_log.erl:1906]: Should we explain why we use this exotic way of converting a string into a binary?

File: src/osiris_log.erl [hunk 2]
```diff
@@ -1903,7 +1903,7 @@ update_retention(Retention,
 evaluate_retention(Dir, Specs) when is_list(Dir) ->
     % convert to binary for faster operations later
     % mostly in segment_from_index_file/1
-    evaluate_retention(list_to_binary(Dir), Specs);
+    evaluate_retention(unicode:characters_to_binary(Dir), Specs);
```
Comment (ansd) [inline src/osiris_log.erl:1906]: You always convert to `unicode` = `utf8` even if the VM is running in `latin1` mode. I think it's probably fine here. Just want to add a note that
https://github.com/erlang/otp/blob/a1f235fc05653fa524fc7c4a352398af6ce0b364/lib/stdlib/src/filename.erl#L845-L851 uses `file:native_name_encoding()` as `OutEncoding`.

**Comment by ansd** [comment]
https://github.com/rabbitmq/osiris/blob/5ff651339dfab5c46d8e8e720ecef469344a94c4/src/osiris_util.erl#L271
I assume we don't support Unicode in the stream name, correct?

Changes make sense to me.
With this branch I also cannot see anymore any formatter crashes that I used to see with the stream test included in https://github.com/rabbitmq/rabbitmq-server/pull/6684

**Review by ansd — APPROVED** [review]

**Comment by kjnilsson** [comment]
> https://github.com/rabbitmq/osiris/blob/5ff651339dfab5c46d8e8e720ecef469344a94c4/src/osiris_util.erl#L271
> 
> 
> I assume we don't support Unicode in the stream name, correct?
> Changes make sense to me. With this branch I also cannot see anymore any formatter crashes that I used to see with the stream test included in [rabbitmq/rabbitmq-server#6684](https://github.com/rabbitmq/rabbitmq-server/pull/6684)

all stream names only contain base64uri compatible characters so this should be fine here.

**Pull request merged**